lib/ocpp/v16/charge_point_impl.cpp

```diff
@@ -2737,6 +2870,248 @@ void ChargePointImpl::handleGetLocalListVersionRequest(ocpp::Call<GetLocalListVe
+        return response;
+    }
+
+    // Find transaction with given id tag
+    std::vector<DisplayMessage> messages;
+    DisplayMessage message;
+    const std::shared_ptr<Transaction> t = this->transaction_handler->get_transaction_from_id_tag(id_token.value());
+    std::string session_id = id_token.value();
+    if (t != nullptr) {
+        session_id = t->get_session_id();
+    } else {
+        EVLOG_error << "Set user price: could not get session id from transaction. Token id is sent instead.";
+    }
+
+    message.transaction_id = session_id;
+
+    if (data.contains("priceText")) {
+        message.message.message = data.at("priceText");
+        if (this->configuration->getLanguage().has_value()) {
+            message.message.language = this->configuration->getLanguage().value();
+        }
+    }
+
+    messages.push_back(message);
+
+    if (this->configuration->getCustomMultiLanguageMessagesEnabled() && data.contains("priceTextExtra") &&
+        data.at("priceTextExtra").is_array()) {
+        for (const json& j : data.at("priceTextExtra")) {
+            DisplayMessage display_message;
+            display_message.transaction_id = session_id;
+            if (j.contains("format")) {
+                display_message.message.message_format =
+                    v201::conversions::string_to_message_format_enum(j.at("format"));
+            }
+
+            if (j.contains("language")) {
+                display_message.message.language = j.at("language");
+            }
+
+            if (j.contains("content")) {
+                display_message.message.message = j.at("content");
+            }
+
+            messages.push_back(display_message);
+        }
+    }
+
+    response = this->set_display_message_callback(messages);
+    return response;
+}
+
+DataTransferResponse ChargePointImpl::handle_set_session_cost(const std::string& type,
+                                                              const std::optional<std::string>& message) {
+    DataTransferResponse response;
+    response.status = DataTransferStatus::Rejected;
+
+    if (!message.has_value()) {
+        EVLOG_error << "Data transfer for RunningCost / FinalCost does not contain any data.";
+        return response;
+    }
+
+    if (session_cost_callback == nullptr) {
+        EVLOG_error << "Received data transfer for " << type << ", but now callback is registered.";
+        return response;
+    }
+
+    RunningCost cost;
+    json data;
+
+    std::optional<std::string> trigger_meter_value_at_time;
+    std::optional<double> trigger_meter_value_at_energy_kwh;
+    std::optional<double> trigger_meter_value_at_power_kw;
+    std::optional<std::vector<ChargePointStatus>> trigger_meter_value_at_chargepoint_status;
+
+    try {
+        data = json::parse(message.value());
+        cost = data;
+
+        // Libocpp will handle those triggers, so they are not stored in the 'RunningCost' object.
+        if (data.contains("triggerMeterValue")) {
```

Comment:
No, I did not add this to the 'cost' object because it is not sent over the interface (because there is no need to send it to other parts of the charging station as libocpp will take care of it).

Reply:
'Solved' it by making a new object here.